### PR TITLE
enable worker graceful shutdown

### PR DESCRIPTION
**Pull request by yux0**
This is a draft version of worker graceful shutdown.
Tests will be provided laster.

### CODE REVIEW

File: internal/internal_workers_test.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ func (s *WorkersTestSuite) TestActivityWorker() {
 	hostEnv := getHostEnvironment()
 	hostEnv.addActivity(a.ActivityType().Name, a)
 	activityWorker := newActivityWorker(
-		s.service, domain, executionParameters, overrides, hostEnv,
+		s.service, domain, executionParameters, overrides, hostEnv, make(chan struct{}),
```
Comment (mfateev) [inline internal/internal_workers_test.go]: We have to add ability to test the worker shutdown notification through ActivityTestEnvironment.

File: internal/activity.go [hunk 2]
```diff
@@ -206,6 +206,22 @@ func GetActivityMetricsScope(ctx context.Context) tally.Scope {
 	return env.metricsScope
 }
 
+// GetWorkerShutdownChannel returns a channel to indicate if the worker is stopped.
```
[the remark this answers is not in the document]
Comment (yux0) [reply internal/activity.go]: Updated the comment. I screwed up my repo and created a new PR: https://github.com/uber-go/cadence-client/pull/677/

File: internal/internal_activity.go [hunk 3]
```diff
@@ -134,9 +134,10 @@ type (
 )
 
 const (
-	activityEnvContextKey          contextKey = "activityEnv"
-	activityOptionsContextKey      contextKey = "activityOptions"
-	localActivityOptionsContextKey contextKey = "localActivityOptions"
+	activityEnvContextKey           contextKey = "activityEnv"
+	activityOptionsContextKey       contextKey = "activityOptions"
+	localActivityOptionsContextKey  contextKey = "localActivityOptions"
+	workerShutdownChannelContextKey contextKey = "workerShutdownChannel"
```
[the remark this answers is not in the document]
Comment (yux0) [reply internal/internal_activity.go]: Make sense. Move it to activity env. Please see: https://github.com/uber-go/cadence-client/pull/677/

File: internal/internal_worker.go [hunk 4]
```diff
@@ -149,6 +149,9 @@ type (
 		// Context to store user provided key/value pairs
 		UserContext context.Context
 
+		// Context cancel function to cancel user context
```
[the remark this answers is not in the document]
Comment (yux0) [reply internal/internal_worker.go]: User context is the background activity context

File: internal/internal_worker_base.go [hunk 5]
```diff
@@ -320,9 +326,26 @@ func (bw *baseWorker) Stop() {
 	// TODO: The poll is longer than wait time, we need some way to hard terminate the
 	// poll routines.
```
[the remark this answers is not in the document]
Comment (yux0) [reply internal/internal_worker_base.go]: Removed this TODO

File: internal/internal_worker_base.go [hunk 6]
```diff
@@ -320,9 +326,26 @@ func (bw *baseWorker) Stop() {
...
+
+	// Close activity channel
+	if bw.activityShutdownCh != nil {
+		close(bw.activityShutdownCh)
+	}
+
 	if success := awaitWaitGroup(&bw.shutdownWG, 2*time.Second); !success {
 		traceLog(func() {
-			bw.logger.Info("Worker timed out on waiting for shutdown.")
+			bw.logger.Info("Activity cancellation timed out.")
```
[the remark this answers is not in the document]
Comment (yux0) [reply internal/internal_worker_base.go]: Reorder the steps. 
1. stop polling and outstanding polling
2. close worker stop channel.
3. wait for the specific timeout
4. cancel context

File: internal/internal_workers_test.go (hunk 1 above)
Comment (yux0) [reply internal/internal_workers_test.go]: Will do in https://github.com/uber-go/cadence-client/pull/677/

**Pull request closed without merging**